pandas/util/_decorators.py

```diff
@@ -268,17 +270,27 @@ def decorator(func: F) -> F:
         def wrapper(*args, **kwargs) -> Callable:
             return func(*args, **kwargs)
 
-        templates = [func.__doc__ if func.__doc__ else ""]
+        # collecting docstring and docstring templates
+        docstring_components: List[Union[str, Callable]] = []
+        if func.__doc__:
+            docstring_components.append(dedent(func.__doc__))
+
         for arg in args:
-            if isinstance(arg, str):
-                templates.append(arg)
-            elif hasattr(arg, "_docstr_template"):
-                templates.append(arg._docstr_template)  # type: ignore
-            elif arg.__doc__:
-                templates.append(arg.__doc__)
-
-        wrapper._docstr_template = "".join(dedent(t) for t in templates)  # type: ignore
-        wrapper.__doc__ = wrapper._docstr_template.format(**kwargs)  # type: ignore
+            if hasattr(arg, "_docstring_components"):
+                docstring_components.extend(arg._docstring_components)  # type: ignore
+            elif isinstance(arg, str) or arg.__doc__:
+                docstring_components.append(arg)
+
+        # formatting templates and concatenating docstring
+        wrapper.__doc__ = "".join(
+            [
+                arg.format(**kwargs) if isinstance(arg, str) else dedent(arg.__doc__)
```

Comment:
The decorator can be used like this:
```python
@doc
def foo():
    pass

@doc(foo)
def bar():
    pass
```
If that happens, in this line `arg` will be the function `foo`, and `arg.__doc__` will be None. When `dedent` is called with `None` as parameter we get:
```
>>> dedent(None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mgarcia/miniconda3/lib/python3.7/textwrap.py", line 430, in dedent
    text = _whitespace_only_re.sub('', text)
TypeError: expected string or bytes-like object
```
And mypy is warning you about it. There are different ways of fixing this. May be this?
```suggestion
                arg.format(**kwargs) if isinstance(arg, str) else dedent(arg.__doc__ or "")
```

Reply:
That make sense.

In line 281 and 282, there are codes to make sure `arg.__doc__` won't be `None`.
```python
elif isinstance(arg, str) or arg.__doc__:
    docstring_components.append(arg)
```

In this case, maybe we can use `# type: ignore`?

Reply:
I think that code ensures `_docstring_components` is not set to `None` (in the parent). And the exception won't happen as far as `_docstring_components` is always set with the `@doc` decorator. But technically speaking there is no guarantee that something like this can be done:
```python
def foo():
    pass
foo._docstring_components = [None]

@doc(foo)
def bar():
    pass
```

So, my preference is an actual fix, like the one I proposed, than a typing `ignore`.

Reply:
> I think that code ensures `_docstring_components` is not set to `None` (in the parent). And the exception won't happen as far as `_docstring_components` is always set with the `@doc` decorator. But technically speaking there is no guarantee that something like this can be done:
> 
> ```python
> def foo():
>     pass
> foo._docstring_components = [None]
> 
> @doc(foo)
> def bar():
>     pass
> ```
> 
> So, my preference is an actual fix, like the one I proposed, than a typing `ignore`.

That is a good point! Error fixed. Thanks!